### PR TITLE
agent: fixes the permissions of PID 1's STDIO

### PR DESCRIPTION
**Pull request by lifupan**
Fix the permissions of PID 1's STDIO within the container to
the specified user.

The ownership needs to match because it is created outside of the
container and needs to be localized.

Fixes: #1022

Signed-off-by: fupan.lfp <fupan.lfp@antfin.com>

### CODE REVIEW

**Review by fidencio — COMMENTED** [review]
It does LGTM, but I'm not familiar with the piece of code.
@bergwolf, @jodh-intel, may I ask for your input here?

Thanks @lifupan!

**Review by devimc — APPROVED** [review]
thanks @lifupan

File: src/agent/rustjail/src/container.rs [hunk 1]
```diff
@@ -639,6 +642,37 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
...
+            // We only change the uid owner (as it is possible for the mount to
+            // prefer a different gid, and there's no reason for us to change it).
+            // The reason why we don't just leave the default uid=X mount setup is
+            // that users expect to be able to actually use their console. Without
+            // this code, you couldn't effectively run as a non-root user inside a
+            // container and also have a console set up.
+            let res = unsafe { libc::fchown(*fd, uid, (0 as libc::gid_t).wrapping_sub(1)) };
+            Errno::result(res)
+                .map(drop)
+                .map_err(|e| anyhow!(e).context("set stdio peerssions failed"))?;
```
Comment (jodh-intel) [inline src/agent/rustjail/src/container.rs:669]: s/peerssions/permissions/

**Review by fidencio — APPROVED** [review]

File: src/agent/rustjail/src/container.rs [hunk 2]
```diff
@@ -639,6 +642,37 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
...
+        if *fd != null_fd {
+            // We only change the uid owner (as it is possible for the mount to
+            // prefer a different gid, and there's no reason for us to change it).
+            // The reason why we don't just leave the default uid=X mount setup is
+            // that users expect to be able to actually use their console. Without
+            // this code, you couldn't effectively run as a non-root user inside a
+            // container and also have a console set up.
+            let res = unsafe { libc::fchown(*fd, uid, (0 as libc::gid_t).wrapping_sub(1)) };
+            Errno::result(res)
+                .map(drop)
```
Comment (jodh-intel) [inline src/agent/rustjail/src/container.rs:668]: Why is this required?

File: src/agent/rustjail/src/container.rs [hunk 3]
```diff
@@ -639,6 +642,37 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
...
+    let null_fd = std::fs::File::open("/dev/null")?.as_raw_fd();
+    let fds = [
+        std::io::stdin().as_raw_fd(),
+        std::io::stdout().as_raw_fd(),
+        std::io::stderr().as_raw_fd(),
+    ];
+
+    for fd in &fds {
+        // Skip chown of /dev/null if it was used as one of the STDIO fds.
+        if *fd != null_fd {
```
Comment (jodh-intel) [inline src/agent/rustjail/src/container.rs:659]: I'd personally invert this to keep the indent minimal:

```rust
 for fd in &fds {
    if *fd == null_fd {
       continue;
   }

   // ...
}
```

**Review by jodh-intel — COMMENTED** [review]
Thanks @lifupan. A few comments.

**Comment by jodh-intel** [comment]
Final comment - is there any way you could add some unit tests for this?

File: src/agent/rustjail/src/container.rs [hunk 4]
```diff
@@ -545,6 +547,8 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
     let uid = Uid::from_raw(guser.uid);
     let gid = Gid::from_raw(guser.gid);
 
+    set_stdio_permissions(guser.uid)?;
```
Comment (bergwolf) [inline src/agent/rustjail/src/container.rs:550]: This assumes that uid exists in the guest. But IMO we should check and create it before using direct, right?

File: src/agent/rustjail/src/container.rs [hunk 5]
```diff
@@ -639,6 +642,37 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
...
+    for fd in &fds {
+        // Skip chown of /dev/null if it was used as one of the STDIO fds.
+        if *fd != null_fd {
+            // We only change the uid owner (as it is possible for the mount to
+            // prefer a different gid, and there's no reason for us to change it).
+            // The reason why we don't just leave the default uid=X mount setup is
+            // that users expect to be able to actually use their console. Without
+            // this code, you couldn't effectively run as a non-root user inside a
+            // container and also have a console set up.
+            let res = unsafe { libc::fchown(*fd, uid, (0 as libc::gid_t).wrapping_sub(1)) };
```
[the remark this answers is not in the document]
Comment (lifupan) [reply src/agent/rustjail/src/container.rs]: Hi @jodh-intel 

I had added some comments for using wrapping_sub, was that clear for you?

File: src/agent/rustjail/src/container.rs (hunk 1 above)
Comment (lifupan) [reply src/agent/rustjail/src/container.rs]: Done

File: src/agent/rustjail/src/container.rs (hunk 2 above)
Comment (lifupan) [reply src/agent/rustjail/src/container.rs]: Removed.

File: src/agent/rustjail/src/container.rs (hunk 3 above)
Comment (lifupan) [reply src/agent/rustjail/src/container.rs]: Done

File: src/agent/rustjail/src/container.rs (hunk 4 above)
Comment (lifupan) [reply src/agent/rustjail/src/container.rs]: If the uid exists, would it be set as 0 by default? I added a check for root user, if the user id is '0', then it would skip the stdio permission set.

**Comment by lifupan** [comment]
> Final comment - is there any way you could add some unit tests for this?

Yes, of course.

File: src/agent/rustjail/src/container.rs (hunk 4 above)
Comment (bergwolf) [reply src/agent/rustjail/src/container.rs]: I was wrong. No need to check it. Kernel doesn't care about `/etc/passwd` at all.

**Review by bergwolf — APPROVED** [review]

**Review by jodh-intel — APPROVED** [review]
Very nice! Thanks @lifupan!

lgtm
